src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs

```diff
@@ -2,58 +2,55 @@
-            : base(threadingContext)
+        public VisualStudioExperimentationService(SVsServiceProvider serviceProvider)
         {
-            object experimentationServiceOpt = null;
-            MethodInfo isCachedFlightEnabledInfo = null;
-            IVsFeatureFlags featureFlags = null;
+            _serviceProvider = (IAsyncServiceProvider)serviceProvider;
+        }
 
-            threadingContext.JoinableTaskFactory.Run(async () =>
+        public async ValueTask<IExperimentationService> GetExperimentationServiceAsync(CancellationToken cancellationToken)
+        {
+            if (_experimentationService is null)
             {
-                try
-                {
-                    featureFlags = (IVsFeatureFlags)await ((IAsyncServiceProvider)serviceProvider).GetServiceAsync(typeof(SVsFeatureFlags)).ConfigureAwait(false);
-                    experimentationServiceOpt = await ((IAsyncServiceProvider)serviceProvider).GetServiceAsync(typeof(SVsExperimentationService)).ConfigureAwait(false);
-                    if (experimentationServiceOpt != null)
-                    {
-                        isCachedFlightEnabledInfo = experimentationServiceOpt.GetType().GetMethod(
-                            "IsCachedFlightEnabled", BindingFlags.Public | BindingFlags.Instance);
-                    }
-                }
-                catch
-                {
-                }
-            });
+                var featureFlags = (IVsFeatureFlags)await _serviceProvider.GetServiceAsync(typeof(SVsFeatureFlags)).ConfigureAwait(false);
+                var telemetryExperimentationService = (Telemetry::IExperimentationService)await _serviceProvider.GetServiceAsync(typeof(SVsExperimentationService)).ConfigureAwait(false);
+                var experimentationService = new ExperimentationService(featureFlags, telemetryExperimentationService);
+                Interlocked.CompareExchange(ref _experimentationService, experimentationService, null);
+            }
 
-            _featureFlags = featureFlags;
-            _experimentationServiceOpt = experimentationServiceOpt;
-            _isCachedFlightEnabledInfo = isCachedFlightEnabledInfo;
+            return _experimentationService;
         }
 
-        public bool IsExperimentEnabled(string experimentName)
+        private class ExperimentationService : IExperimentationService
         {
-            ThisCanBeCalledOnAnyThread();
-            if (_isCachedFlightEnabledInfo != null)
+            private readonly IVsFeatureFlags _featureFlags;
+            private readonly Telemetry::IExperimentationService _experimentationService;
+
+            public ExperimentationService(IVsFeatureFlags featureFlags, Telemetry::IExperimentationService experimentationService)
+            {
+                _featureFlags = featureFlags;
+                _experimentationService = experimentationService;
+            }
+
+            public bool IsExperimentEnabled(string experimentName)
```

Comment:
I would simply hold onto IAsyncServiceProvider here. and then lazily set IVsFeatureFlags and IExperimentationService?

and provide both IsExperimentEnabledAsync and IsExperimentEnabled version.

also, since Task.True and Task.False is cached in framework, IsExpereientEnabledAsync returning either true or false won't cause an allocation. so no reason to use ValueTask.

Reply:
It is not possible to correctly implement this service without any asynchronous methods. I opted to push the asynchronous call early so callers could cache the resulting value and avoid the need to only use IsExperimentEnabledAsync on asynchronous code paths.